### ISSUE TITLE
SQLInsertClause#executeWithKeys leaks cursor when stmt.executeUpdate throws

### ISSUE THREAD

**Issue by jnape** [querydsl/querydsl#2213]
Line **368** of `com.querydsl.sql.dml.SQLInsertClause` calls `stmt.executeUpdate()`. If that throws a `SQLException`, then `stmt` (a `PreparedStatement`) is never closed. 

An example unit test that illustrates this would be:

```java
    @Test
    public void executeWithKeysClosesStatementInLightOfSqlException() throws SQLException {
        Connection conn = createMock(Connection.class);
        PreparedStatement stmt = createMock(PreparedStatement.class);
        expect(conn.prepareStatement(eq("insert into EMPLOYEE\nvalues ()"), aryEq(new String[]{"ID"}))).andReturn(stmt);
        expect(stmt.executeUpdate()).andThrow(new SQLException("expected"));
        stmt.close();
        replayAll();

        QEmployee emp1 = new QEmployee("emp1");
        SQLInsertClause insert = new SQLInsertClause(conn, SQLTemplates.DEFAULT, emp1);
        try {
            insert.executeWithKeys();
            fail("expected QueryException to have been thrown, but wasn't.");
        } catch (QueryException expected) {
            verify(stmt);
        }
    }
```

Attaching the patch file seemed to be the easiest option for everyone, but let me know if you'd prefer a pull request.

[querydsl.txt](https://github.com/querydsl/querydsl/files/1411754/querydsl.txt)

**Comment by Shredder121** [comment]
If you would use the SQLQueryFactory, it should close the Connection for you (since it manages the connections).
If you supply a Connection to a SQLQuery as-is, it will only close it if you registered the SQLCloseListener in the Configuration.
Otherwise, it can't close the Connection, since it's been given as argument, and closing it always would be wrong.

**Comment by jnape** [comment]
Sorry for the long feedback loop here, this fell off my radar.

Hmm, I think maybe I wasn't clear, or else I don't understand your point. Specifically, if you look at the lines I'm referencing, you'll see that on line *362* of `SQLInsertClause`, a method-local `PreparedStatement` is created, then unsafe code is executed. If that code then throws an exception, the `PreparedStatement` leaks, guaranteed. There is absolutely no measure that can be taken from user-code to ensure this reference is closed properly.

Are you suggesting a way to maneuver around this code being executed at all? If so, certainly the library maintainers would still want to fix or deprecate this method, right?

Sorry if there was any confusion in my initial report.

**Comment by jnape** [comment]
Bump.

**Comment by jwgmeligmeyling** [comment]
@jnape is this still on your radar?

**Comment by jnape** [comment]
I wouldn't say it's still on my radar: I locally forked and patched these leaks in query dsl years ago when it became clear this bug was getting no traction. For what it's worth, it certainly seems like it's still pervasive in [`AbstractSQLInsertClause#executeWithKeys` on master](https://github.com/querydsl/querydsl/blob/master/querydsl-sql/src/main/java/com/querydsl/sql/dml/AbstractSQLInsertClause.java#L359).

**Comment by jwgmeligmeyling** [comment]
QueryDSL is under active maintenance again, so if you're local changes are still available somewhere on a fork I'd certainly be willing to have a look and merge them in to the main code base.

**Comment by jnape** [comment]
That's a generous offer, but again, my local fork is from 2017, so it seems like it's pretty far out of date. However, the patch file I attached to the first comment on this issue when I opened it should give you a good sense of exactly what the nature of the bug was and how I went about fixing it.

The long short of it is that there are `PreparedStatement`s that QueryDSL internally requisitions and manages that are never exposed outside of QueryDSL internals, and there was faulty exception handling logic that allowed scenarios like the one I ran into (addressed by the patch file attached to this issue), where a `PreparedStatement` is executed, a subsequent internal QueryDSL code path throws an exception before the statement is closed, and the statement leaks. My patch file addressed the specific place where this actually happened in production for me, and is part of the code patch I applied locally to fix it -- again, it is for the state of master from 2017, so I suspect it's unlikely to cleanly apply at this point.

Given how old this issue is and how far out of sync I am with the current state of QueryDSL, I think I've exhausted my appetite for continuing to attempt to shephard this issue. Hopefully I've provided enough context to help to identify these resource leaks if they do indeed still exist in the current master, but if not, I doubt I'll be able to better communicate the circumstances of their manifestation than I've already feebly attempted.

Cheers, and best of luck with the ongoing development of the very successful project!
